### PR TITLE
docs(blog): add sqlsmith blog by b41sh

### PR DESCRIPTION
**Pull request by PsiACE**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13431)

### CODE REVIEW

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+description: "Discover bugs before your users do! Learn how we implement our SQLsmith, a  randomized SQL test generator, to find 50+ bugs in Databend. Read this blog to see how it works."
+authors:
+  - name: b41sh
+    url: https://github.com/b41sh
+    image_url: https://github.com/b41sh.png
+---
+
+# Leveraging Randomized Testing with SQLsmith to Uncover Bugs
+
+Testing plays a crucial role in the development and maintenance of database systems. It not only verifies correctness of functionality but also uncovers potential issues to ensure performance and stability with each change and release. Databend’s CI already supports various test types including:
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:17]:
```suggestion
Testing plays a crucial role in the development and maintenance of database systems. It not only verifies correctness of functionality but also uncovers potential issues to ensure performance and stability with each change and release. Databend’s CI already supports these test types:
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+- **Unit Tests**: Validate functionality of minimum testable code units like functions and modules. Ensure they execute and return expected results.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:19]:
```suggestion
- **Unit Test**: A unit test validates the functionality of minimum testable code units like functions and modules, and ensure they execute successfully and return the expected results.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+- **SQL Logic Tests**: Verify SQL syntax and logic correctness via test cases. Cover different query scenarios to ensure queries run properly in all situations.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:20]:
```suggestion
- **SQL Logic Test**: A SQL logic test verifies SQL syntax and logic correctness using test cases, covering various scenarios.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+- **Performance Tests**: Validate performance impact of new features and optimizations. Prevent performance regressions.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:21]:
```suggestion
- **Performance Test**: A performance test validates performance impact of new features and optimizations. This helps prevent performance regressions.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+While these tests ensure functional correctness and stability during rapid development, they have limitations. Test SQL queries are mostly handwritten and simple, lacking complex queries seen in real-world scenarios. They have insufficient coverage of edge cases and exceptions. 
+
+## Introducing SQLsmith
+
+[SQLsmith](https://github.com/anse1/sqlsmith) is a randomized SQL query generator that produces a high volume of diverse test cases to simulate real-world variability. Compared to other testing approaches, SQLsmith improves test coverage to uncover more potential issues and bugs.
+
+## Implementing Our SQLsmith
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:29]:
```suggestion
## Implementing SQLsmith for Databend
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+The original SQLsmith for PostgreSQL fuzz testing was inspired by [Csmith](https://github.com/csmith-project/csmith). Now, many known open source databases have their own SQLsmith ports like [CockroachDB](https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/sqlsmith/sqlsmith.go), [TiDB](https://github.com/PingCAP-QE/go-sqlsmith), [RisingWave](https://github.com/risingwavelabs/risingwave/tree/main/src/tests/sqlsmith) etc.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:31]:
```suggestion
The original SQLsmith for PostgreSQL fuzz testing was inspired by [Csmith](https://github.com/csmith-project/csmith). Now, many known open source databases have their own SQLsmith ports like [CockroachDB](https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/sqlsmith/sqlsmith.go), [TiDB](https://github.com/PingCAP-QE/go-sqlsmith), and [RisingWave](https://github.com/risingwavelabs/risingwave/tree/main/src/tests/sqlsmith). 
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
+---
+title: "Leveraging Randomized Testing with SQLsmith to Uncover Bugs"
+date: 2023-10-24
+slug: 2023-10-24-sqlsmith-hero
+cover_url: 'sqlsmith-hero.png'
+image: 'sqlsmith-hero.png'
+tags: ["fuzz testing", "SQLsmith", "test"]
+description: "Discover bugs before your users do! Learn how we implement our SQLsmith, a  randomized SQL test generator, to find 50+ bugs in Databend. Read this blog to see how it works."
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:8]:
```suggestion
description: "Discover bugs before your users do! Learn how we implement our SQLsmith, a randomized SQL test generator, to find 50+ bugs in Databend. Read this blog to see how it works."
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+## Implementing Our SQLsmith
+
+The original SQLsmith for PostgreSQL fuzz testing was inspired by [Csmith](https://github.com/csmith-project/csmith). Now, many known open source databases have their own SQLsmith ports like [CockroachDB](https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/sqlsmith/sqlsmith.go), [TiDB](https://github.com/PingCAP-QE/go-sqlsmith), [RisingWave](https://github.com/risingwavelabs/risingwave/tree/main/src/tests/sqlsmith) etc. 
+
+These open source SQLsmiths use diverse languages (C++, Go, Rust) with different syntax support suitable for their domains. We couldn't directly use them and had to build our own SQLsmith in Rust to fully support Databend's syntax and features.
+
+SQLsmith has three main components:
+
+- **SQL Generator**: Generates abstract syntax trees (ASTs) of various types
+- **SQL Reducer**: Simplifies failing SQLs to isolate bug causes
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:38]:
```suggestion
- **SQL Reducer**: Simplifies failed SQLs to isolate bug causes
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+- **Runner**: Executes SQLs, records errors
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:39]:
```suggestion
- **Runner**: Executes SQLs and records errors 
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+# Leveraging Randomized Testing with SQLsmith to Uncover Bugs
+
+Testing plays a crucial role in the development and maintenance of database systems. It not only verifies correctness of functionality but also uncovers potential issues to ensure performance and stability with each change and release. Databend’s CI already supports various test types including:
+
+- **Unit Tests**: Validate functionality of minimum testable code units like functions and modules. Ensure they execute and return expected results.
+- **SQL Logic Tests**: Verify SQL syntax and logic correctness via test cases. Cover different query scenarios to ensure queries run properly in all situations.
+- **Performance Tests**: Validate performance impact of new features and optimizations. Prevent performance regressions.
+
+While these tests ensure functional correctness and stability during rapid development, they have limitations. Test SQL queries are mostly handwritten and simple, lacking complex queries seen in real-world scenarios. They have insufficient coverage of edge cases and exceptions.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:23]:
```suggestion
While these tests ensure functional correctness and stability during rapid development, they have limitations. Handwritten SQL queries are often simplistic and lack coverage of complex real-world scenarios, edge cases, and exceptions.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+- **SQL Generator**: Generates abstract syntax trees (ASTs) of various types
+- **SQL Reducer**: Simplifies failing SQLs to isolate bug causes
+- **Runner**: Executes SQLs, records errors 
+
+## SQL Generator
+
+The SQL Generator randomly generates ASTs including:
+
+- **Data Types**: Basic types, Nested types like `Array`, `Map`, `Tuple` etc.
+- **DDL**: `CREATE`, `ALTER`, `DROP TABLE` etc. to create and modify test tables.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:46]:
```suggestion
- **DDLs**: `CREATE`, `ALTER`, `DROP TABLE` etc. to create and modify test tables.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+- **DML**: `INSERT`, `UPDATE`, `DELETE`, `MERGE` etc. to populate test data.
+- **Query**: `WITH`, `SELECT`, `JOIN`, `SubQuery`, `ORDER BY` etc. `WITH` and `SubQuery` can generate complex queries.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:48]:
```suggestion
- **Queries**: `WITH`, `SELECT`, `JOIN`, `SubQuery`, `ORDER BY` etc. `WITH` and `SubQuery` can generate complex queries.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+- **Expression**: `Column`, `Literal`, `Scalar Function` and `Aggregate Function` etc. Expressions can be nested.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:49]:
```suggestion
- **Expressions**: `Column`, `Literal`, `Scalar Function` and `Aggregate Function` etc. Expressions can be nested.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+By randomly generating AST components, it covers all possible SQL syntax. Controlled recursion also creates complex nested SQLs to find obscure bugs. But depth is capped to avoid unexecutable SQLs.
+
+## SQL Reducer
+
+Since generated SQLs can be very complex with the bug trigger buried, simplifying the original SQL makes bug isolation easier.
+
+The SQL Reducer iteratively removes AST components like `WITH`, `SubQuery`, `Expression` etc. If the simplified SQL still reproduces the bug, it's used. Else original SQL is retained. Ultimately, it outputs the smallest SQL to reproduce the bug by traversing all AST parts.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:57]:
```suggestion
The SQL Reducer iteratively removes AST components like `WITH`, `SubQuery`, `Expression` etc. If the simplified SQL continues to trigger the bug, it is adopted; otherwise, the original SQL is preserved. In the end, it generates the most concise SQL to reproduce the bug by systematically examining all AST components.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+## Runner 
+
+The SQL Runner loops generating test SQLs, executes them using Databend, and processes failures as:
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:61]:
```suggestion
The SQL Runner executes test SQL queries using Databend, and in case of failure, it proceeds with the following error-handling steps:
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+- Check for expected errors like syntax, semantics etc.
+- Verify if it's a known issue or unimplemented feature.
+- Call the SQL Reducer to generate the smallest reproducible SQL.
+- Log errors and simplified SQLs.
+
+The Runner is integrated into Databend's CI, executing on every release to log issues for further analysis.
+
+## Impact
+
+Thus far, SQLsmith has uncovered [50+ bugs](https://github.com/datafuselabs/databend/issues?q=is%3Aissue+label%3Afound-by-sqlsmith) in Databend after a month of running, including:
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:72]:
```suggestion
After a month of running, SQLsmith has uncovered [50+ bugs](https://github.com/datafuselabs/databend/issues?q=is%3Aissue+label%3Afound-by-sqlsmith) in Databend, including:
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  
+- Internal logic errors (17 bugs)
+- Invalid function/expression checks (12 bugs)  
+- Missing semantic checks (9 bugs)
+- Improper `unwrap`/`unreachable` handling (7 bugs) 
+- Parser failures (3 bugs)
+- Failed casts across types (4 bugs)
+- Parquet I/O errors (1 bug)
+
+## Key Takeaways
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:82]:
```suggestion
## Key Lessons Learned
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+## What's Next
+
+SQLsmith generates most common SQLs and integrates with CI. Upcoming improvements include:
+
+- Supporting more SQL features like `UNION`, **Computed Column** etc.
+- More configurations like expression nesting depths, query complexity.
+- Improved SQL simplification.
+- Optimized query execution and analysis.
+
+SQLsmith is a powerful tool to discover hidden bugs and improve Databend's stability and reliability. We look forward to its evolving role.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:104]:
```suggestion

```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+   - For `Int` parameters, check for large number values.
+   - When supporting any parameter type, consider unusual types like `Null`, `EmptyArray`, `Bitmap` etc.
+   - When only supporting specific types, see if other types can auto-convert or check and return errors early.
+2. Use unwrap judiciously. Explicitly handle `Result` and `Option` instead of assuming success.
+3. Understand SQL semantic rules during development e.g. constraints around `GROUP BY`, `ORDER BY` etc. Perform semantic checks in the Binder phase to prevent runtime errors.
+4. Add more unit test cases for critical modules to prevent bugs from internal logic errors.
+
+## What's Next
+
+SQLsmith generates most common SQLs and integrates with CI. Upcoming improvements include:
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:97]: SQLsmith helps enhance Databend's stability and reliability by uncovering hidden issues. It generates common SQL queries, integrates with CI, and upcoming improvements include:

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+- Supporting more SQL features like `UNION`, **Computed Column** etc.
+- More configurations like expression nesting depths, query complexity.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:100]:
```suggestion
- Introduce more configurations like expression nesting depths, query complexity.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+- Improved SQL simplification.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:101]:
```suggestion
- Improve SQL Reducer.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+- Optimized query execution and analysis.
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:102]:
```suggestion
- Optimize query executions and result analysis.
```

File: website/blog/2023-10-24-sqlsmith-hero.md [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+   - For `Int` parameters, check for large number values.
+   - When supporting any parameter type, consider unusual types like `Null`, `EmptyArray`, `Bitmap` etc.
+   - When only supporting specific types, see if other types can auto-convert or check and return errors early.
+2. Use unwrap judiciously. Explicitly handle `Result` and `Option` instead of assuming success.
+3. Understand SQL semantic rules during development e.g. constraints around `GROUP BY`, `ORDER BY` etc. Perform semantic checks in the Binder phase to prevent runtime errors.
+4. Add more unit test cases for critical modules to prevent bugs from internal logic errors.
+
+## What's Next
+
+SQLsmith generates most common SQLs and integrates with CI. Upcoming improvements include:
```
Comment (soyeric128) [inline website/blog/2023-10-24-sqlsmith-hero.md:97]:
```suggestion
SQLsmith helps enhance Databend's stability and reliability by uncovering hidden issues. It generates common SQL queries, integrates with CI, and upcoming improvements include:
```

**Review by soyeric128 — APPROVED** [review]
